Definition support from Java to properties for ConfigProperty/name

angelozerr opened the pull request:
Definition support from Java to properties for ConfigProperty/name

Fixes #88

Signed-off-by: azerr <azerr@redhat.com>

angelozerr commented:
This PR requires the vscode-microprofile PR https://github.com/redhat-developer/vscode-microprofile/pull/29

datho7561 commented:
Looks good so far

angelozerr commented:
This PR provides

* improve the hover of Configproperty/name with hyperlink for properties file

![image](https://user-images.githubusercontent.com/1932211/100888164-1969fa00-34b6-11eb-8d49-b297f5c682fe.png)

* the definition support:
  * from Java to Java. The usecase if for fault tolerance

![FallbackDefinitionDemo](https://user-images.githubusercontent.com/1932211/100888477-736abf80-34b6-11eb-975c-a08076758154.gif)

  * from Java to properties file

![ConfigPropertyDefinitionDemo](https://user-images.githubusercontent.com/1932211/100888731-b6c52e00-34b6-11eb-8ac5-d96b7ac5c94d.gif)

Please note for yaml file usecase (from Java to yaml file), it opens just the application.yaml file without selecting the property.

datho7561 commented on microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/core/project/MicroProfileConfigPropertyInformation.java:
```diff
@@ -29,6 +29,7 @@
 	private final String propertyNameWithProfile;
 	private final String value;
 	private final String configFileName;
```
If you have the URI, you can probably get derive the file name from the URI. eg. `Paths.get(sourceConfigFileUri).getName()`. The file name is just for display purposes.

angelozerr replied:
sourceConfigFileUri can be null, it's the reason why I did like that

datho7561 reviewed (changes requested):
Around halfway through the code changes. The tests are failing

datho7561 commented on microprofile.ls/org.eclipse.lsp4mp.ls/src/main/java/org/eclipse/lsp4mp/utils/URIUtils.java:
```diff
@@ -0,0 +1,141 @@
+	 * @param baos    the {@code ByteArrayOutputStream} to be copied into a String
+	 * @param charset the {@link Charset} to use to decode the bytes
+	 * @return the String that has been copied to (possibly empty)
+	 * @since 5.2.6
```
Remove, not relevent to this project

angelozerr replied:
fixed

datho7561 commented:
Some of the tests are failing, and more tests need to be added, but the code looks pretty good, and from my manual testing it is working well.

datho7561 commented on microprofile.jdt/org.eclipse.lsp4mp.jdt.test/src/main/java/org/eclipse/lsp4mp/jdt/core/faulttolerance/MicroProfileFaultToleranceJavaDefinitionTest.java:
```diff
@@ -0,0 +1,51 @@
+		// @Fallback(fallbackMethod = "b|bb") --> public String bbb()
+		assertJavaDefinitions(p(21, 33), javaFileUri, utils, //
+				def(r(21, 32, 35), javaFileUri, r(26, 18, 21)));
+	}
```
I think a test with an empty `@Fallback(|)` would be helpful to double check that no NPE are thrown in this case.

angelozerr replied:
The test was added

datho7561 commented:
The diagnostics appear twice for fallbackMethod:

![Screenshot from 2020-12-03 14-13-25](https://user-images.githubusercontent.com/22376627/101076736-d0777b80-3571-11eb-9b90-9d1bf5c80521.png)

Also, in this case, it links `fallbackMethod` to the `Data` class

datho7561 commented:
It will also link to methods in another class, but this doesn't seem to be allowed (for Quarkus at least)
![LinksToMethodInDifferentClass](https://user-images.githubusercontent.com/22376627/101077422-ce61ec80-3572-11eb-874a-2de65677f8c1.gif)

angelozerr commented:
> The diagnostics appear twice for fallbackMethod:

fixed with test (problem came from inner class)

> It will also link to methods in another class, but this doesn't seem to be allowed (for Quarkus at least)

fixed with test (problem came from inner class)

angelozerr commented on microprofile.jdt/org.eclipse.lsp4mp.jdt.test/projects/maven/microprofile-fault-tolerance/src/main/java/org/acme/FaultTolerantResourceWithInnerClass.java:
```diff
@@ -0,0 +1,47 @@
+package org.acme;
```
See `class Data {` at the end of the file which causes the bugs that you reported.

datho7561 replied:
Its present in both `FaultTolerantResource.java` and `FaultTolerantResourceWithInnerClass.java`

angelozerr replied:
Good catch! I removed this class.
